integration test fails and leaks a JVM on java 9EA

This is separate from the commons lang issue affecting 'deb' (#14700).
This one impacts 'zip'

```
=======================================
Elasticsearch Build Hamster says Hello!
=======================================
  Gradle Version : 2.8
  JDK Version    : 1.9.0-ea-b91 (Oracle Corporation)
  OS Info        : Linux 4.2.0-16-generic (amd64)
:core:compileJava
Note: Some input files use or override a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
Note: Some input files use unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.
:core:processResources
:core:classes
:core:jar
:distribution:zip:buildZip
:rest-api-spec:compileJava UP-TO-DATE
:rest-api-spec:processResources
:rest-api-spec:classes
:rest-api-spec:jar
:distribution:zip:processTestResources UP-TO-DATE
:distribution:zip:copyRestSpec
:test-framework:compileJava
Note: Some input files use or override a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
Note: Some input files use unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.
:test-framework:processResources
:test-framework:classes
:test-framework:jar
:distribution:zip:compileTestJava
:distribution:zip:testClasses
:distribution:zip:integTest#clean
:distribution:zip:integTest#checkPrevious SKIPPED
:distribution:zip:integTest#stopPrevious SKIPPED
:distribution:zip:integTest#extract
:distribution:zip:integTest#configure
:distribution:zip:integTest#start
Java HotSpot(TM) 64-Bit Server VM warning: Option UseParNewGC was deprecated in version 9.0 and will likely be removed in a future release.

[2015-11-12 01:25:37,075][INFO ][node                     ] [Living Tribunal] version[3.0.0-SNAPSHOT], pid[26943], build[c2aec53/2015-11-12_01:25:06]
[2015-11-12 01:25:37,075][INFO ][node                     ] [Living Tribunal] initializing ...
[2015-11-12 01:25:37,117][INFO ][plugins                  ] [Living Tribunal] loaded [], sites []
[2015-11-12 01:25:37,132][INFO ][env                      ] [Living Tribunal] using [1] data paths, mounts [[/ (/dev/sdb1)]], net usable_space [85.7gb], net total_space [229.1gb], spins? [no], types [ext4]
[2015-11-12 01:25:38,411][INFO ][node                     ] [Living Tribunal] initialized
[2015-11-12 01:25:38,411][INFO ][node                     ] [Living Tribunal] starting ...
[2015-11-12 01:25:38,518][WARN ][common.network           ] [Living Tribunal] publish host: [_local_] resolves to multiple addresses, auto-selecting {127.0.0.1} as single publish address
[2015-11-12 01:25:38,518][INFO ][transport                ] [Living Tribunal] publish_address {127.0.0.1:9500}, bound_addresses {[::1]:9500}, {127.0.0.1:9500}
[2015-11-12 01:25:38,526][INFO ][discovery                ] [Living Tribunal] distribution_zip_integTest/QGrcE43yRtiRtIpn7RpmXQ
[2015-11-12 01:25:41,547][INFO ][cluster.service          ] [Living Tribunal] new_master {Living Tribunal}{QGrcE43yRtiRtIpn7RpmXQ}{127.0.0.1}{127.0.0.1:9500}[testattr=>test], reason: zen-disco-join(elected_as_master, [0] joins received)
[2015-11-12 01:25:41,595][WARN ][common.network           ] [Living Tribunal] publish host: [_local_] resolves to multiple addresses, auto-selecting {127.0.0.1} as single publish address
[2015-11-12 01:25:41,595][INFO ][http                     ] [Living Tribunal] publish_address {127.0.0.1:9400}, bound_addresses {[::1]:9400}, {127.0.0.1:9400}
[2015-11-12 01:25:41,596][INFO ][node                     ] [Living Tribunal] started
:distribution:zip:integTest#start FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':distribution:zip:integTest#start'.
> Failed to start elasticsearch

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.

BUILD FAILED

Total time: 39.31 secs
rmuir@beast:~/workspace/elasticsearch/distribution/zip$ jps
27032 Jps
21882 org.eclipse.equinox.launcher_1.3.100.v20150511-1540.jar
26943 Elasticsearch

```

Elasticsearch in fact started up fine, and you can use it. So there are two bugs, probably the first one to fix is the JVM leak, since i've hit that a few times this week already.

--- a/buildSrc/src/main/groovy/org/elasticsearch/gradle/test/ClusterFormationTasks.groovy
+++ b/buildSrc/src/main/groovy/org/elasticsearch/gradle/test/ClusterFormationTasks.groovy
@@ -18,21 +18,16 @@
  */
 package org.elasticsearch.gradle.test
 
-import org.gradle.internal.jvm.Jvm
-
-import java.nio.file.Paths
-
+import org.apache.tools.ant.DefaultLogger
 import org.apache.tools.ant.taskdefs.condition.Os
 import org.elasticsearch.gradle.VersionProperties
-import org.gradle.api.DefaultTask
-import org.gradle.api.GradleException
-import org.gradle.api.InvalidUserDataException
-import org.gradle.api.Project
-import org.gradle.api.Task
+import org.gradle.api.*
 import org.gradle.api.file.FileCollection
 import org.gradle.api.tasks.Copy
 import org.gradle.api.tasks.Delete
 import org.gradle.api.tasks.Exec
+
+import java.nio.file.Paths
 
 /**
  * A helper for creating tasks to build a cluster that is used by a task, and tear down the cluster when the task is finished.
@@ -216,23 +211,57 @@ class ClusterFormationTasks {
             'JAVA_HOME' : project.javaHome,
             'ES_GC_OPTS': config.jvmArgs
         ]
-        List esProps = config.systemProperties.collect { key, value -> "-D${key}=${value}" }
+        List<String> esProps = config.systemProperties.collect { key, value -> "-D${key}=${value}" }
         for (Map.Entry<String, String> property : System.properties.entrySet()) {
             if (property.getKey().startsWith('es.')) {
                 esProps.add("-D${property.getKey()}=${property.getValue()}")
             }
         }
 
-        Closure esPostStartActions = { ant, logger ->
-            ant.waitfor(maxwait: '30', maxwaitunit: 'second', checkevery: '500', checkeveryunit: 'millisecond', timeoutproperty: "failed${name.capitalize()}") {
+        String executable
+        List<String> esArgs = []
+        if (Os.isFamily(Os.FAMILY_WINDOWS)) {
+            executable = 'cmd'
+            esArgs.add('/C')
+            esArgs.add('call')
+        } else {
+            executable = 'sh'
+        }
+        // running with cmd on windows will look for this with the .bat extension
+        esArgs.add(new File(home, 'bin/elasticsearch').toString())
+
+        // this closure is converted into ant nodes by groovy's AntBuilder
+        Closure antRunner = {
+            exec(executable: executable, spawn: config.daemonize, dir: cwd, taskname: 'elasticsearch') {
+                esEnv.each { key, value -> env(key: key, value: value) }
+                (esArgs + esProps).each { arg(value: it) }
+            }
+            waitfor(maxwait: '30', maxwaitunit: 'second', checkevery: '500', checkeveryunit: 'millisecond', timeoutproperty: "failed${name}") {
                 and {
                     resourceexists {
-                        file file: pidFile.toString()
+                        file(file: pidFile.toString())
                     }
                     http(url: "http://localhost:${config.httpPort}")
                 }
             }
+        }
+
+        // this closure is the actual code to run elasticsearch
+        Closure elasticsearchRunner = {
+            ByteArrayOutputStream buffer = new ByteArrayOutputStream()
+            if (logger.isInfoEnabled() || config.daemonize == false) {
+                // run with piping streams directly out (even stderr to stdout since gradle would capture it)
+                runAntCommand(project, antRunner, System.out, System.err)
+            } else {
+                // buffer the output, we may not need to print it
+                PrintStream captureStream = new PrintStream(buffer, true, "UTF-8")
+                runAntCommand(project, antRunner, captureStream, captureStream)
+            }
+
             if (ant.properties.containsKey("failed${name}".toString())) {
+                // the waitfor failed, so dump any output we got (may be empty if info logging, but that is ok)
+                logger.error(buffer.toString('UTF-8'))
+                // also dump the cluster's log file, it may be useful
                 File logFile = new File(home, "logs/${clusterName}.log")
                 if (logFile.exists()) {
                     logFile.eachLine { line -> logger.error(line) }
@@ -240,35 +269,10 @@ class ClusterFormationTasks {
                 throw new GradleException('Failed to start elasticsearch')
             }
         }
-        File esScript = new File(home, 'bin/elasticsearch')
-
-        if (Os.isFamily(Os.FAMILY_WINDOWS)) {
-            // elasticsearch.bat is spawned as it has no daemon mode
-            return project.tasks.create(name: name, type: DefaultTask, dependsOn: setup) << {
-                // Fall back to Ant exec task as Gradle Exec task does not support spawning yet
-                ant.exec(executable: 'cmd', spawn: config.daemonize, dir: cwd) {
-                    esEnv.each { key, value -> env(key: key, value: value) }
-                    (['/C', 'call', esScript] + esProps).each { arg(value: it) }
-                }
-                esPostStartActions(ant, logger)
-            }
-        } else {
-            List esExecutable = [esScript]
-            if(config.daemonize) {
-                esExecutable.add("-d")
-            }
-
-            return project.tasks.create(name: name, type: Exec, dependsOn: setup) {
-                workingDir cwd
-                executable 'sh'
-                args esExecutable
-                args esProps
-                environment esEnv
-                doLast {
-                    esPostStartActions(ant, logger)
-                }
-            }
-        }
+        
+        Task start = project.tasks.create(name: name, type: DefaultTask, dependsOn: setup)
+        start.doLast(elasticsearchRunner)
+        return start
     }
 
     /** Adds a task to check if the process with the given pidfile is actually elasticsearch */
@@ -346,5 +350,17 @@ class ClusterFormationTasks {
 
     static File pidFile(File dir) {
         return new File(dir, 'es.pid')
+    }
+
+    /** Runs an ant command, sending output to the given out and error streams */
+    static void runAntCommand(Project project, Closure command, PrintStream outputStream, PrintStream errorStream) {
+        DefaultLogger listener = new DefaultLogger(
+                errorPrintStream: errorStream,
+                outputPrintStream: outputStream,
+                messageOutputLevel: org.apache.tools.ant.Project.MSG_INFO)
+
+        project.ant.project.addBuildListener(listener)
+        project.configure(project.ant, command)
+        project.ant.project.removeBuildListener(listener)
     }
 }

